clippy_lints/src/mem_replace.rs

```diff
@@ -194,64 +193,36 @@ fn check_replace_with_uninit(cx: &LateContext<'_>, src: &Expr<'_>, dest: &Expr<'
     }
 }
 
-/// Returns true if the `def_id` associated with the `path` is recognized as a "default-equivalent"
-/// constructor from the std library
-fn is_default_equivalent_ctor(cx: &LateContext<'_>, def_id: DefId, path: &QPath<'_>) -> bool {
-    let std_types_symbols = &[
-        sym::string_type,
-        sym::vec_type,
-        sym::vecdeque_type,
-        sym::LinkedList,
-        sym::hashmap_type,
-        sym::BTreeMap,
-        sym::hashset_type,
-        sym::BTreeSet,
-        sym::BinaryHeap,
-    ];
-
-    if let QPath::TypeRelative(_, method) = path {
-        if method.ident.name == sym::new {
-            if let Some(impl_did) = cx.tcx.impl_of_method(def_id) {
-                if let Some(adt) = cx.tcx.type_of(impl_did).ty_adt_def() {
-                    return std_types_symbols
-                        .iter()
-                        .any(|&symbol| cx.tcx.is_diagnostic_item(symbol, adt.did));
-                }
-            }
+fn check_replace_with_default(cx: &LateContext<'_>, src: &Expr<'_>, dest: &Expr<'_>, expr_span: Span) {
+    // disable lint for primitives
+    if let ExprKind::Lit(_) = src.kind {
+        return;
+    }
```

Comment:
I done it but not sure if it is the best line we can draw. Specially I don't like disabling lint for slices and `&str`.

Reply:
Hmm good point. `&str`, slices and tuples are considered "aggregate data types", so we could lint those? I think that also removes "recursively" from the picture so just `is_non_aggregate_primitive`.

@rust-lang/clippy any thoughts? We currently only lint things like `mem::replace(x, Vec::new())` where there is some constructor call or `Default::default()`. I think suggesting `mem::take` instead of `mem::replace(x, false)` or `0` is weird so I suggested in my previous comment above to exclude all primitives. But now I'm thinking to exclude "non-aggregate primitives".

Reply:
i think that could work!

Reply:
@HKalbasi does that sound good to you?

Reply:
Yes it is good, and `if let ExprKind::Lit(_) = src.kind` was doing the same but it depend on the implementation of `is_default_equivalent`. I didn't find `is_non_aggregate_primitive`. If it doesn't exist with another name, I should add this if I understand correctly in `ty.rs`:
```rs
pub fn is_non_aggregate_primitive_type(ty: Ty<'_>) -> bool {
    matches!(ty.kind(), ty::Bool | ty::Char | ty::Int(_) | ty::Uint(_) | ty::Float(_))
}
```
Is it covers all non aggregate primitives? And another option is putting `matches!` directly in if condition.

Reply:
> Yes it is good, and `if let ExprKind::Lit(_) = src.kind` was doing the same but it depend on the implementation of `is_default_equivalent`.

I think `ExprKind::Lit` would exclude string literals.

> If it doesn't exist with another name, I should add this if I understand correctly in `ty.rs`:

Looks good. I think it should be a util.